### PR TITLE
manual: add "declarative package management" section

### PR DESCRIPTION
**Pull request by matthewbauer**
This section gives some details on how to setup an "environment"
without having to go through NixOS (although it could be used there
too). I’ve tried to make it straightforward and have a kind of
"tutorial" feel. Not sure if that’s appropriate for the manual, so any
recommended changes would be helpful.

Also the section might not be totally appropriate, but this is a fairly general use case, I think.

### CODE REVIEW

**Comment by Ericson2314** [comment]
Night there be a wiki migration issue this fixes?

File: doc/configuration.xml [hunk 1]
```diff
@@ -243,5 +243,218 @@ set of packages.
...
+export MANPATH=$HOME/.nix-profile/share/man:/nix/var/nix/profiles/default/share/man:/usr/share/man
+    '';
+    myPackages = pkgs.buildEnv {
+      name = "my-packages";
+      paths = [
+        (runCommand "profile" {} ''
+mkdir -p $out/etc/profile.d
+cp ${myProfile} $out/etc/profile.d/my-profile.sh
+        '')
+        aspell
```
Comment (LnL7) [inline doc/configuration.xml]: I think it might be better to stick to 2 or 3 packages for the examples.

**Comment by vcunat** [comment]
Just for record, I personally prefer using simple attrmaps instead of `buildEnv`; I find that more flexible and better working together with `nix-env`.  Still, in general I do find it useful to document such howto/cookbook/best-practice stuff.

**Comment by matthewbauer** [comment]
> Just for record, I personally prefer using simple attrmaps instead of buildEnv

Do you have an example of your setup? My understanding was you could never get man pages/info pages setup correctly without ```extraOutputsToInstall```.

**Comment by vcunat** [comment]
Correct; I *do* consider #24717 a bug.  It's still possible to map over the custom set and amend `meta.outputsToInstall`, as an example workaround.

File: doc/configuration.xml [hunk 2]
```diff
@@ -243,5 +243,218 @@ set of packages.
...
+      paths = [ aspell bc coreutils gdb ffmpeg nixUnstable emscripten jq nox silver-searcher ];
+      pathsToLink = [ "/share" "/bin" ];
+    };
+  };
+}
+    </screen>
+
+    <para>
+      <literal>pathsToLink</literal> tells Nixpkgs to only link the paths listed
+      which gets rid of the extra stuff in the profile.
```
Comment (0xABAB) [inline doc/configuration.xml]: `not linking the extra stuff (like <insert concrete non-matching file path>) in the profile.` (feel free to modify slightly if what I am saying is not correct, the point I want to make is that it's vague)

File: doc/configuration.xml [hunk 3]
```diff
@@ -243,5 +243,218 @@ set of packages.
...
+      <filename>/bin</filename> and <filename>/share</filename> are good
+      defaults for a user environment, getting rid of the clutter. If you are
```
Comment (0xABAB) [inline doc/configuration.xml]: `defaults for a clutter free user environment` is shorter.

File: doc/configuration.xml [hunk 4]
```diff
@@ -243,5 +243,218 @@ set of packages.
...
+    </para>
+
+  </section>
+
+  <section xml:id="sec-getting-documentation">
+    <title>Getting documentation</title>
+
+    <para>
+      After building that new environment, look through
+      <filename>~/.nix-profile</filename> to make sure everything is there that
```
Comment (0xABAB) [inline doc/configuration.xml]: -`that we wanted`

File: doc/configuration.xml [hunk 5]
```diff
@@ -243,5 +243,218 @@ set of packages.
...
+      we wanted. Discerning readers will note that some files are missing. Look
```
Comment (0xABAB) [inline doc/configuration.xml]: A fairly long paragraph to make a simple point. Spending words to tell the user is super smart, might be attractive to the latent narcissists , but I count them as waste.

File: doc/configuration.xml [hunk 6]
```diff
@@ -243,5 +243,218 @@ set of packages.
...
+      paths = [ aspell bc coreutils ffmpeg nixUnstable emscripten jq nox silver-searcher ];
+      pathsToLink = [ "/share/man" "/share/doc" /bin" ];
+      extraOutputsToInstall = [ "man" "doc" ];
+    };
+  };
+}
+    </screen>
+
+    <para>
+      This provides us with some useful documentation for using our packages.
```
Comment (0xABAB) [inline doc/configuration.xml]: This line doesn't add anything.

File: doc/configuration.xml [hunk 7]
```diff
@@ -243,5 +243,218 @@ set of packages.
...
+      However, if we actually want those manpages to be detected by man, we need
```
Comment (0xABAB) [inline doc/configuration.xml]: To make man detect our written manpages...

File: doc/configuration.xml [hunk 8]
```diff
@@ -243,5 +243,218 @@ set of packages.
...
+      to set up our environment. This can also be managed within Nix
```
Comment (0xABAB) [inline doc/configuration.xml]: Overloading the word environment makes it harder to read.

File: doc/configuration.xml [hunk 9]
```diff
@@ -243,5 +243,218 @@ set of packages.
...
+
+    <para>
+      For this to work fully, you must also have this script sourced when you
+      are logged in. Try adding something like this to your
+      <filename>~/.profile</filename> file:
+    </para>
+
+    <screen>
+#!/bin/sh
+if [ -d $HOME/.nix-profile/etc/profile.d ]; then
```
Comment (0xABAB) [inline doc/configuration.xml]: Not quoting $HOME makes baby Jezus cry. It's better to replace it with `~`.

File: doc/configuration.xml [hunk 10]
```diff
@@ -243,5 +243,218 @@ set of packages.
...
+  for i in $HOME/.nix-profile/etc/profile.d/*.sh; do
```
Comment (0xABAB) [inline doc/configuration.xml]: Same

File: doc/configuration.xml [hunk 11]
```diff
@@ -243,5 +243,218 @@ set of packages.
...
+    if [ -r $i ]; then
```
Comment (0xABAB) [inline doc/configuration.xml]: Quotes

File: doc/configuration.xml [hunk 12]
```diff
@@ -243,5 +243,218 @@ set of packages.
...
+      . $i
```
Comment (0xABAB) [inline doc/configuration.xml]: Quotes

File: doc/configuration.xml [hunk 13]
```diff
@@ -243,5 +243,218 @@ set of packages.
...
+    fi
+  done
+fi
+    </screen>
+
+    <para>
+      Now just run <literal>source $HOME/.profile</literal> and you can starting
```
Comment (0xABAB) [inline doc/configuration.xml]: `~`

File: doc/configuration.xml [hunk 14]
```diff
@@ -243,5 +243,218 @@ set of packages.
...
+      loading man pages from your environent.
+    </para>
+
+  </section>
+    
+  <section xml:id="sec-gnu-info-setup">
+    <title>GNU info setup</title>
+
+    <para>
+      Configuring GNU info is a little bit trickier than man pages. To work
```
Comment (0xABAB) [inline doc/configuration.xml]: Qualifications about difficulty are a waste of words. 

To let GNU Info display our <info pages/documentation/whatever> it needs this information indexed in a database, which is done as followed:

File: doc/configuration.xml [hunk 15]
```diff
@@ -243,5 +243,218 @@ set of packages.
...
+        silver-searcher
+        texinfoInteractive
+      ];
+      pathsToLink = [ "/share/man" "/share/doc" "/share/info" "/bin" "/etc" ];
+      extraOutputsToInstall = [ "man" "doc" "info" ];
+      postBuild = ''
+          if [ -x $out/bin/install-info -a -w $out/share/info ]; then
+            shopt -s nullglob
+            for i in $out/share/info/*.info $out/share/info/*.info.gz; do
+                $out/bin/install-info $i $out/share/info/dir
```
Comment (0xABAB) [inline doc/configuration.xml]: Quotes

File: doc/configuration.xml [hunk 16]
```diff
@@ -243,5 +243,218 @@ set of packages.
...
+            done
+          fi
+      '';
+    };
+  };
+}
+    </screen>
+
+    <para>
+      <literal>postBuild</literal> tells Nixpkgs to run a command after building
```
Comment (0xABAB) [inline doc/configuration.xml]: Nixpkgs doesn't really run anything. Please rewrite to be correct.

File: doc/configuration.xml [hunk 17]
```diff
@@ -243,5 +243,218 @@ set of packages.
...
+      the environment. In this case, <literal>install-info</literal> adds the
+      installed info pages to <literal>dir</literal> which is GNU info's default
```
Comment (0xABAB) [inline doc/configuration.xml]: GNU Info (see https://www.gnu.org/software/texinfo/manual/info-stnd/info-stnd.html)

File: doc/configuration.xml [hunk 18]
```diff
@@ -243,5 +243,218 @@ set of packages.
...
+      root node. Note that <literal>texinfoInteractive</literal> is added to the
+      environment to give the <literal>install-info</literal> command.
```
Comment (0xABAB) [inline doc/configuration.xml]: `to give` is unspecific language.

File: doc/configuration.xml [hunk 19]
```diff
@@ -243,5 +243,218 @@ set of packages.
...
+  <section xml:id="sec-getting-documentation">
+    <title>Getting documentation</title>
+
+    <para>
+      After building that new environment, look through
+      <filename>~/.nix-profile</filename> to make sure everything is there that
+      we wanted. Discerning readers will note that some files are missing. Look
+      inside <filename>~/.nix-profile/share/man/man1/</filename> to verify this.
+      There are no man pages for any of the Nix tools! This is because some
+      packages like Nix have multiple outputs for things like documentation (see
```
Comment (0xABAB) [inline doc/configuration.xml]: DocBook is a hyperlinked format, so just link to the section.

File: doc/configuration.xml [hunk 20]
```diff
@@ -243,5 +243,218 @@ set of packages.
 
 </section>
 
+<section xml:id="sec-declarative-package-management">
+  <title>Declarative Package Management</title>
+
+  <section xml:id="sec-building-environment">
+    <title>Build an environment</title>
+
```
Comment (0xABAB) [inline doc/configuration.xml]: Please define what is understood by an environment.

**Review by 0xABAB — CHANGES_REQUESTED** [review]
Thanks for taking the time to write documentation, but I have quite some comments.

**Comment by grahamc** [comment]
Merging, this is better than what is there. Thank you.

**Pull request merged**